llm_on_ray/finetune/DataPreprocess.py

```diff
@@ -0,0 +1,224 @@
+                    instruction,
+                    re.DOTALL,
+                )
+                convs_tokens = [
+                    tokenizer.tokenize(conv) + tokenizer.tokenize("\n") for conv in convs
+                ]
+                header_tokens = tokenizer.tokenize(header) + tokenizer.tokenize("\n")
+                max_input = max_source_length - len(header_tokens) - len(assistant_tokens)
+                truncated_convs = truncate_sequences(convs_tokens, max_input)
+                if len(truncated_convs) == 0:
+                    truncated_convs = [convs_tokens[-1][: max_input - 3] + convs_tokens[-1][-3:]]
+
+                prompt_tokens = [header_tokens] + truncated_convs + [assistant_tokens]
+                prompt_ids = [
+                    tokenizer.convert_tokens_to_ids(prompt_token) for prompt_token in prompt_tokens
+                ]
+                prompt_ids = list(chain(*prompt_ids))
+
+                resp_ids = tokenizer.convert_tokens_to_ids(tokenizer.tokenize(response.strip()))
+                # keep last and eos_id
+                max_resp = max_seq_length - len(prompt_ids) - 1
+
+                # truncating response
+                if len(resp_ids) > max_resp:
+                    if truncation_side == "right":
+                        resp_ids = resp_ids[: max_resp - 1] + resp_ids[-1:]
+                    else:
+                        resp_ids = resp_ids[-max_resp:]
+
+                # masking
+                input_ids = prompt_ids + resp_ids + [tokenizer.eos_token_id]
+                if mask_input:
+                    labels = [IGNORE_INDEX] * len(prompt_ids) + resp_ids + [tokenizer.eos_token_id]
+                elif mask_response:
+                    labels = prompt_ids + [IGNORE_INDEX] * len(resp_ids) + [tokenizer.eos_token_id]
+                else:
+                    labels = input_ids
+
+                # padding
+                input_len = len(input_ids)
+                pad_len = max_seq_length - input_len
+                if padding_side == "right":
+                    input_ids = input_ids + [tokenizer.eos_token_id] * pad_len
+                    labels = labels + [IGNORE_INDEX] * pad_len
+                    attention_mask = [1] * input_len + [0] * pad_len
+                else:
+                    input_ids = [tokenizer.eos_token_id] * pad_len + input_ids
+                    labels = [IGNORE_INDEX] * pad_len + labels
+                    attention_mask = [0] * pad_len + [1] * input_len
+
+                assert len(input_ids) == max_seq_length
+                assert len(prompt_ids) <= max_source_length
+                assert len(labels) == len(input_ids) == len(attention_mask)
+
+                examples["input_ids"].append(torch.tensor(input_ids))
+                examples["labels"].append(labels)
+                examples["attention_mask"].append(attention_mask)
+
+            return examples
+
+        def preprocess_function_with_tokenizer(examples):
```

Comment:
this function is not clear too much compared with `preprocess_function_with_tokenize`
will get confused from function name